drivers/i2c/i2c_omap.c

```diff
@@ -0,0 +1,921 @@
+#define I2C_WE_ARDY_WE BIT(2)  /* Register access ready wakeup */
+#define I2C_WE_NACK_WE BIT(1)  /* No acknowledgment wakeup */
+#define I2C_WE_AL_WE   BIT(0)  /* Arbitration lost wakeup */
+
+#define I2C_WE_ALL                                                                                 \
+	(I2C_WE_XDR_WE | I2C_WE_RDR_WE | I2C_WE_AAS_WE | I2C_WE_BF_WE | I2C_WE_STC_WE |            \
+	 I2C_WE_GC_WE | I2C_WE_DRDY_WE | I2C_WE_ARDY_WE | I2C_WE_NACK_WE | I2C_WE_AL_WE)
+
+/* I2C Buffer Configuration Register (I2C_BUF): */
+#define I2C_BUF_RXFIF_CLR BIT(14) /* RX FIFO Clear */
+#define I2C_BUF_TXFIF_CLR BIT(6)  /* TX FIFO Clear */
+
+/* I2C Status Register (I2C_STAT): */
+#define I2C_STAT_XDR  BIT(14) /* TX Buffer draining */
+#define I2C_STAT_RDR  BIT(13) /* RX Buffer draining */
+#define I2C_STAT_BB   BIT(12) /* Bus busy */
+#define I2C_STAT_ROVR BIT(11) /* Receive overrun */
+#define I2C_STAT_XUDF BIT(10) /* Transmit underflow */
+#define I2C_STAT_AAS  BIT(9)  /* Address as target */
+#define I2C_STAT_XRDY BIT(4)  /* Transmit data ready */
+#define I2C_STAT_RRDY BIT(3)  /* Receive data ready */
+#define I2C_STAT_ARDY BIT(2)  /* Register access ready */
+#define I2C_STAT_NACK BIT(1)  /* No ack interrupt enable */
+#define I2C_STAT_AL   BIT(0)  /* Arbitration lost */
+
+/* I2C System Test Register (I2C_SYSTEST): */
+#define I2C_SYSTEST_ST_EN       BIT(15)   /* System test enable */
+#define I2C_SYSTEST_FREE        BIT(14)   /* Free running mode */
+#define I2C_SYSTEST_TMODE_MASK  (3 << 12) /* Test mode select mask */
+#define I2C_SYSTEST_TMODE_SHIFT (12)      /* Test mode select shift */
+
+/* Functional mode */
+#define I2C_SYSTEST_SCL_I_FUNC BIT(8) /* SCL line input value */
+#define I2C_SYSTEST_SDA_I_FUNC BIT(6) /* SDA line input value */
+
+/* SDA/SCL IO mode */
+#define I2C_SYSTEST_SCL_I BIT(3) /* SCL line sense in */
+#define I2C_SYSTEST_SCL_O BIT(2) /* SCL line drive out */
+#define I2C_SYSTEST_SDA_I BIT(1) /* SDA line sense in */
+#define I2C_SYSTEST_SDA_O BIT(0) /* SDA line drive out */
+
+struct i2c_omap_cfg {
+	mem_addr_t base;
+	uint32_t irq;
+	uint32_t speed;
+	init_func_t init_func;
+};
+
+enum i2c_omap_speed {
+	I2C_OMAP_SPEED_STANDARD,
+	I2C_OMAP_SPEED_FAST,
+	I2C_OMAP_SPEED_FAST_PLUS,
+};
+
+struct i2c_omap_speed_config {
+	uint32_t pscstate;
+	uint32_t scllstate;
+	uint32_t sclhstate;
+};
+
+struct i2c_omap_data {
+	enum i2c_omap_speed speed;
+	struct i2c_omap_speed_config speed_config;
+	struct i2c_msg current_msg;
+	struct k_sem cmd_complete;
+	int cmd_err;
+	bool receiver;
+	bool bb_valid;
+};
+/**
+ * Writes a value to the register of the OMAP I2C controller.
+ *
+ * @param cfg The configuration structure for the OMAP I2C controller.
+ * @param reg The register to write the value to.
+ * @param value The value to write to the register.
+ */
+static inline void i2c_omap_reg_set(const struct i2c_omap_cfg *cfg, uint32_t reg, uint32_t value)
+{
+	sys_write16(value, cfg->base + reg);
+}
```

Comment:
These are all 32-bit registers as per the TRM, even though the upper bits are reserved , that does not mean you should access it as 16-bit registers?

Reply:
Understood!
I will change it to sys32.